crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -250,69 +261,209 @@ pub fn impl_param_set(_input: TokenStream) -> TokenStream {
-        if attrs.ignore {
-            ignored_fields.push(field.ident.as_ref().unwrap());
-            ignored_field_types.push(&field.ty);
-        } else {
-            field_locals.push(format_ident!("f{i}"));
-            let i = Index::from(i);
-            fields.push(
-                field
-                    .ident
-                    .as_ref()
-                    .map(|f| quote! { #f })
-                    .unwrap_or_else(|| quote! { #i }),
-            );
-            field_types.push(&field.ty);
+    for (i, field) in fields.iter().enumerate() {
+        let mut field_attrs = SystemParamAttributes::default();
+        for attr in &field.attrs {
+            let Some(attr_ident) = attr.path.get_ident() else { continue; };
+            if attr_ident == SYSTEM_PARAM_ATTRIBUTE_NAME {
+                if field_attrs.usage.is_none() {
+                    field_attrs.usage =
+                        Some(match attr.parse_args_with(SystemParamFieldUsage::parse) {
+                            Ok(o) => o,
+                            Err(e) => return e.into_compile_error().into(),
+                        });
+                } else {
+                    return syn::Error::new(
+                        attr.span(),
+                        "Multiple `system_param` field attributes found.",
+                    )
+                    .into_compile_error()
+                    .into();
+                }
+            }
+        }
+
+        match field_attrs
+            .usage
+            .unwrap_or_else(|| fallibility.clone().into())
+        {
+            SystemParamFieldUsage::Ignore => {
+                ignored_fields.push(match field.ident.as_ref() {
+                    Some(s) => s,
+                    None => {
+                        return syn::Error::new(field.span(), "Field lacks an identifier.")
+                            .into_compile_error()
+                            .into()
+                    }
+                });
+                ignored_field_types.push(&field.ty);
+            }
+            field_fallibility => {
+                let ident = format_ident!("f{i}");
+                let i = Index::from(i);
+                let ty = &field.ty;
+                field_idents.push(
+                    field
+                        .ident
+                        .as_ref()
+                        .map(|f| quote! { #f })
+                        .unwrap_or_else(|| quote! { #i }),
+                );
+                field_getters.push(match fallibility {
+					SystemParamStructUsage::Infallible => match field_fallibility {
+						SystemParamFieldUsage::Infallible => quote! { #ident },
+						SystemParamFieldUsage::Optional => quote! { #ident.expect("Optional system param was not infallible!") },
+						SystemParamFieldUsage::Resultful => quote! { #ident.expect("Resultful system param was not infallible!") },
+						_ => unreachable!(),
+					},
```

Comment:
If I understand this correctly, this means that if a custom `SystemParam` has a field with incorrect fallibility, it will panic at runtime instead of failing at compile time. I don't think this is an acceptable failure mode.

Reply:
Here is an example of that:
```rust
#[derive(SystemParam)] // By default, it's infallible, which means it must return or panic
struct MyParam<'w> {
  #[system_param(optional)]
  foo: Res<'w, SomeResource>,
}
```
What it is doing is getting the param as `Option<Res<SomeResource>>` instead of `Res<SomeResource>` and unwrapping it. The only difference from not getting it as `Res<SomeResource>` is that if it panics, it will panic here instead of from inside the `SystemParam` implementation for `Res<SomeResource>`.

Essentially, you would never want to write this, because it's pointless. However, I kept it in for consistency and for usability in macros (it makes it easier to write macros when there are less edge cases to handle). Otherwise, there aren't any downsides.